### PR TITLE
feat(middleware): Implement request short-circuiting via Response.complete

### LINKED ISSUE #1296: feat: Add a way for Middleware process_request() to indicate whether routing should proceed or the response should just be rendered immediately
We can either set an attribute on `resp`, or simply support returning a truthy value to indicate whether or not the routing should be short-circuited. Other ideas?

Should we also do this for process_resource()?

### PR DESCRIPTION
**Pull request by kgriffs**
Closes #1296

### CODE REVIEW

File: docs/api/middleware.rst [hunk 1]
```diff
@@ -112,13 +112,43 @@ like this::
...
+exception, it will be processed according to the exception type. If
 the type matches a registered error handler, that handler will
 be invoked and then the framework will begin to unwind the
 stack, skipping any lower layers. The error handler may itself
-raise an instance of HTTPError, in which case the framework
-will use the latter exception to update the *resp* object.
+raise an instance of :class:`~.HTTPError` or :class:`~.HTTPStatus`, in
+which case the framework will use the latter exception to update the
+*resp* object.
+
```
Comment (kgriffs) [inline docs/api/middleware.rst]: While I was updating the docs for this PR, I made a couple of additional updates that aren't directly related, but I decided to include them in the same PR in the interest of getting things merged more quickly for 2.0.

File: CHANGES.rst [hunk 2]
```diff
@@ -32,6 +32,8 @@ New & Improved
 --------------
 
 - Added a new ``headers`` property to the ``Response`` class.
+- Added a new ``complete`` property to the ``Response`` class. This can be used
```
Comment (vytas7) [inline CHANGES.rst]: Change not mirrored to `2.0.0` changelog yet.

File: falcon/api.py [hunk 3]
```diff
@@ -220,21 +220,24 @@ def __call__(self, env, start_response):  # noqa: C901
                 if self._independent_middleware:
                     for process_request in mw_req_stack:
                         process_request(req, resp)
+                        if resp.complete:
+                            break
                 else:
                     for process_request, process_response in mw_req_stack:
-                        if process_request:
+                        if process_request and not resp.complete:
```
Comment (vytas7) [inline falcon/api.py]: Unless I missed something I'm getting an impression that this combination has no test coverage:
* `independent_middleware=False`
* skipping `process_request` due to `resp.complete`

File: falcon/api.py [hunk 4]
```diff
@@ -244,12 +247,16 @@ def __call__(self, env, start_response):  # noqa: C901
                     # route, a default responder is returned and the
                     # resource is None. In that case, we skip the
                     # resource middleware methods.
-                    if resource is not None:
+                    if resource is not None and not resp.complete:
```
Comment (vytas7) [inline falcon/api.py]: Same here, I could remove `and not resp.complete` and still get everything to pass.

File: falcon/api.py (hunk 3 above)
Comment (kgriffs) [reply falcon/api.py]: Yikes, good catch! Covered now.

File: falcon/api.py (hunk 4 above)
Comment (kgriffs) [reply falcon/api.py]: In this case, when `resp.complete == True`, `resource` will always be `None`, which is why the check is superfluous. I went ahead and removed it and updated the comment.

File: tox.ini [hunk 5]
```diff
@@ -88,7 +88,7 @@ deps = {[with-debug-tools]deps}
        jsonschema
 
 [testenv:py3_debug]
-basepython = python3.6
+basepython = python3.7
```
Comment (vytas7) [inline tox.ini]: Good catch, I must have missed these :+1: 

While we are at it, could we also change the new `check_vendored` environment which was developed on a parallel track?

**Pull request merged**